base/strings/annotated.jl

```diff
@@ -386,3 +386,132 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+
+function write(io::AnnotatedIOBuffer, astr::Union{AnnotatedString, SubString{<:AnnotatedString}})
+    astr = AnnotatedString(astr)
+    offset = position(io.io)
+    if !eof(io)
+        # If we are overwriting an existing span in the AnnotatedIOBuffer,
+        # clear out any overlapping pre-existing annotations.
+        span = offset+1:offset+ncodeunits(astr)
+        filter!(((range, _),) -> first(range) < first(span) || last(range) > last(span), io.annotations)
+        extras = Tuple{UnitRange{Int}, Pair{Symbol, Any}}[]
+        for i in eachindex(io.annotations)
+            range, annot = io.annotations[i]
+            # Test for partial overlap
+            if first(range) <= first(span) <= last(range) || first(range) <= last(span) <= last(range)
+                io.annotations[i] = (if first(range) < first(span)
+                                         first(range):first(span)-1
+                                     else last(span)+1:last(range) end, annot)
+                # If `span` fits exactly within `range`, then we've only copied over
+                # the beginning overhang, but also need to conserve the end overhang.
+                if first(range) < first(span) && last(span) < last(range)
+                    push!(extras, (last(span)+1:last(range), annot))
+                end
+            end
+            # Insert any extra entries in the appropriate position
+            for entry in extras
+                indices = searchsorted(io.annotations, (first(entry),), by=first)
+                splice!(io.annotations, indices, Tuple{UnitRange{Int}, Pair{Symbol, Any}}[entry])
+            end
+        end
+        # Fill in the annotations from `astr`, accounting for `offset`
+        for (region, annot) in astr.annotations
+            region = first(region)+offset:last(region)+offset
+            indices = searchsorted(io.annotations, (region,), by=first)
+            splice!(io.annotations, indices, Tuple{UnitRange{Int}, Pair{Symbol, Any}}[(region, annot)])
+        end
+    else
+        # Fill in the annotations from `astr`, accounting for `offset`
+        for (region, annot) in astr.annotations
+            region = first(region)+offset:last(region)+offset
+            push!(io.annotations, (region, annot))
+        end
+    end
+    write(io.io, String(astr))
+end
+
+write(io::AnnotatedIOBuffer, c::AnnotatedChar) = write(io, AnnotatedString(c))
+write(io::AnnotatedIOBuffer, x::AbstractString) = write(io.io, x)
+write(io::AnnotatedIOBuffer, s::Union{SubString{String}, String}) = write(io.io, s)
+write(io::AnnotatedIOBuffer, b::UInt8) = write(io.io, b)
+
+function write(dest::AnnotatedIOBuffer, src::AnnotatedIOBuffer)
+    destpos = position(dest)
+    srcpos = position(src)
+    offset = destpos - srcpos
+    nb = write(dest.io, src.io)
+    # Clear out annotations that only start after `position(dest)`,
+    filter!(((region, _),) -> first(region) <= destpos, dest.annotations)
```

Comment:
Should this put back afterwards any annotations that end after the `position(destpos)`, so that it only overwrites the mid-portion of the annotations? (In lisp, I would say to use `separate!` instead of `filter!`, but I don't know if that exists in Julia)

Reply:
```suggestion
    # Preserve any annotations that end after `position(dest)`
    destpos2 = position(dest)
    endannotations = filter(((region, _),) -> last(region) >= destpos2, dest.annotations)
    # Truncate those existing annotation to start after `destpos2`
    map!(((region, annot),) -> (max(first(region), destpos2):last(region), annot),
         endannotations, endannotations)
    # Clear out annotations that only start after `position(dest)`,
    filter!(((region, _),) -> first(region) <= destpos, dest.annotations)
```

Reply:
Ah yes, that would seem sensible. That said, there's a little more too it since annotations that go after the end of the write may need to have their region trimmed at the start too. This logic is already in the `write(io::AnnotatedIOBuffer, astr::Union{AnnotatedString, SubString{<:AnnotatedString}})` method, so I've extracted the logic into a helper function `_clear_annotations_in_region!` that can be used here too.